---
title: "gperez_Assign8"
author: "Gera"
date: "9/13/2019"
output: html_document
---

This is an exercise for KEGG Gene Set Enrichment Analysis and Visualization of RNA-Seq data

setwd("/Users/gerardoperez/Documents/shell/Bi623/Assign8")

```{r engine = 'bash', eval = FALSE}
scp /projects/bgmp/shared/Bi623/assign8/* .

```

1. Load the libraries gage, gageData, and pathview, and read in the edgeR differential gene expression results for pipefish


```{r}
library(gage)
library(gageData)
library(pathview)

# import file into R for analysis
kegg_pouch <- read.delim("pouch_RNAseq.tsv", sep = "\t",    
                         stringsAsFactors=FALSE)

head(kegg_pouch)

```

2. Set up and perform GSEA for all KEGG pathways using gage


```{r}
#Loads the KEGG orthology data from the gageData database

data(kegg.sets.ko)
data(sigmet.idx.ko)
kegg.sets.ko <- kegg.sets.ko[sigmet.idx.ko]
head(kegg.sets.ko, 3)
class(kegg.sets.ko)
```

What class of object is kegg.sets.ko, and what kind of information does it contain?

**The class object for kegg.sets.ko is a list. It contains kegg pathway entries with its orthology.** 


```{r}
#Constructs a vector of log2 fold change values and use the names() function to name each component with a ko ID
#Important: will allow to do GSEA for all of the KEGG pathways, based on log2 fold change, per-gene measure of differential expression.

pouch_foldchanges <- kegg_pouch$logFC
names(pouch_foldchanges) <- kegg_pouch$ko_ID
head(pouch_foldchanges)

#Runs gage() on our vector of log2 fold change values to test, for each KEGG pathway, enrichment for genes with extreme values.

pouch_test <- gage(pouch_foldchanges, gsets=kegg.sets.ko, 
                   same.dir=FALSE)

#Performs a “two-tailed” test, meaning a significantly enriched KEGG pathway can be defined by extreme fold changes in either direction.
#Uses lapply() on the head() function to look at the top entries of the two elements in our list

lapply(pouch_test, head)
head(pouch_test$greater, 30)

#Returns only those pathways such that our false discovery rate (FDR) is controlled at 0.1
pouch_test_FDR_0.1<-subset(pouch_test$greater, pouch_test$greater[,4]<0.1)


```

Which KEGG pathways are enriched for genes with exceptional pregnancy-specific gene expression? 

**ko04060 Cytokine-cytokine receptor interaction**

**ko04080 Neuroactive ligand-receptor interaction** 

**ko04020 Calcium signaling pathway**  

**ko04610 Complement and coagulation cascades**    

Which two of these are related to the immune system?

**ko04060 Cytokine-cytokine receptor interaction**

**ko04610 Complement and coagulation cascades**  

3. Visualize pregnancy fold change magnitudes for all genes in the “coagulation and complement” KEGG pathway



```{r}
# Gets the  pathway and its ko ID from the gage() output by simple indexing and the substr() function.

pouch_pathways <- rownames(pouch_test$greater)[4]
pouch_ids <- substr(pouch_pathways, start=1, stop=7)
pouch_ids

#Draws the pathway with a color scale that reflects log2 fold change for each gene.

#The pathview() function will write 3 files to your working directory. “ko04610.pathview.png” is the one you want to have a look at.

pathview(gene.data=pouch_foldchanges, 
         pathway.id=pouch_ids, species="ko", new.signature=FALSE,
         trans.fun = list(gene = NULL, cpd = NULL), 
         low = list(gene = "green", cpd = "blue"), 
         mid = list(gene = "yellow", cpd = "gray"), 
         high = list(gene = "red", cpd = "yellow"), na.col = "transparent")


```

What does the plot tell you about components of this pathway and male pregnancy?

**The plot shows that the Extrinsic pathway (tissue damage), and Intrinsic pathway (contact with damaged vessel) are upregulated in pregnant male pipefish,**

4. Visualize multi-genic expression patterns in pregnant and non-pregnant pouch tissues using non-metric multidimensional scaling


```{r}
library(vegan)
library(MASS)

# import file into R for analysis
pipe_TMMvals <- read.csv("pouch_TMM_values.csv", head = TRUE,   
                           row.names = 1)
dim(pipe_TMMvals)
pipe_TMMvals <- t(pipe_TMMvals)
#head(pipe_TMMvals)

```


What happened to our data frame after using the t() function?

**The data went from 15253 observations of 12 variables to a large matrix of 183036 elements that outputs the first column of the original data frame after executing it with the head function. It made the matrix transpose and made the rows into columns suggesting that this is the transpose function**




```{r}
#Computes a dissimilarity matrix (distances between pipefish samples in “transcript space”) using the default for vegdist(), which is a metric called Bray-Curtis dissimilarity.

pipe.dis <- vegdist(pipe_TMMvals)
                    

```

What do you think is meant by “convergence”? Think about how nmds works, and consult the literature, if necessary.


**A best fit that has minimum stress levels for both axis. Reducing the stress levels of distances 15,252 dimensions into 2 dimensions by distances.**

```{r}
# Fits of the new, 2-D distances to the original, 15,252-D distances
# Visualizes a stress plot (or Shepard plot), which plots the 2-D versus 15,252-D distance ranks.

pipe.mds0 <- isoMDS(pipe.dis, k=2)
stressplot(pipe.mds0, pipe.dis)

# constructs a data frame that links the pouch sample IDs with pregnancy status.

Targets <- as.data.frame(rownames(pipe_TMMvals))
Targets$PregStat <- factor(c(rep("preg",6),rep("nonpreg",6)))
colnames(Targets) <- c("ID","PregStat")

#Defines the ordination plotting parameters
# ordiplot() function will produce an empty space

par(mgp = c(2.5, 1, 0))
preg=as.character(Targets$PregStat)

fig <- ordiplot(pipe.mds0, main="Brood Pouches in Transcript Space",
                ylab="nMDS Dimension 2", xlab="nMDS Dimension 1", 
                font.lab=2, font.axis=2, cex.axis=.7, type="none", cex.main=1,
                xlim=(c(-.2,0.2)))


#Adds “confidence ellipses” and the individual samples as points using the ordiellipse() and points() functions

ordiellipse(pipe.mds0,groups=preg,label=FALSE, lwd=2,  
              show.groups=preg[1:6], col="darkorange4",
            draw="lines")
ordiellipse(pipe.mds0,groups=preg,label=FALSE, lwd=2, 
              show.groups=preg[7:12], col="dodgerblue4",
            draw="lines")

points(fig, "sites", pch=c(rep(19,6),rep(15,6)), 
         col=c(rep("darkorange3",6),rep("cornflowerblue",6)), cex=1.5)

```

What does this plot tell you about the brood pouch transcriptomes profiled in this study?

**Transcriptiome clusters separately by pregnancy status. Also there is more Beta diversity in pregnant males compared to non-pregnant males.**

```{r}
#Repeats the ordination (starting with the isoMDS() call), but this time generates 3 nMDS dimensions as opposed to 2.

pipe.mds1 <- isoMDS(pipe.dis, k=3)
stressplot(pipe.mds1, pipe.dis)

Targets <- as.data.frame(rownames(pipe_TMMvals))
Targets$PregStat <- factor(c(rep("preg",6),rep("nonpreg",6)))
colnames(Targets) <- c("ID","PregStat")

par(mgp = c(2.5, 1, 0))
preg=as.character(Targets$PregStat)

#head(pipe.mds1$points)

fig2 <- ordiplot(pipe.mds1$points[,2:3], main="Brood Pouches in Transcript Space",
                ylab="nMDS Dimension 3", xlab="nMDS Dimension 2", 
                font.lab=2, font.axis=2, cex.axis=.7, type="none", cex.main=1,
                xlim=(c(-.2,0.2)))

ordiellipse(pipe.mds1$points[,2:3],groups=preg,label=FALSE, lwd=2,  
            show.groups=preg[1:6], col="darkorange4",
            draw="lines")
ordiellipse(pipe.mds1$points[,2:3],groups=preg,label=FALSE, lwd=2, 
            show.groups=preg[7:12], col="dodgerblue4",
            draw="lines")

points(fig2, "sites", pch=c(rep(19,6),rep(15,6)), 
       col=c(rep("darkorange3",6),rep("cornflowerblue",6)), cex=1.5)


```

5.Permutational Multivariate Analysis of Variance (perMANOVA) to test for multivariate transcriptional differences between pregnant and non-pregnant males


```{r}
#Defines a dataframe with Targets data frame to execute the adonis() function, which runs the perMANOVA,

otu.env <- Targets
adonis(pipe.dis ~ PregStat, otu.env, perm=999)

```



Based on the output of adonis(), do we see a significant effect of pregnancy status?

**26.18 % is the proportion of total dissimilarity explained by pregnancy status. 26.18 % variation is explained by the pregnancy status of the individuals. This has a p-value of 0.003 which is pretty significant.**


6. Constructing a heatmap with clustering dendrograms for Coagulation and Complement Cascade KEGG pathway genes.

```{r}
library(gplots)
library(RColorBrewer)
library(dendextend)

# import file into R for analysis
pouch_compcoag <- read.delim("CompCoag_pouch_multivar.tsv", sep = "\t", 
                             row.names = 1, header = F)

colnames(pouch_compcoag)<- c("KO","name","P9","P8","P7","P6","P5","P3",
                                        "NP11","NP10","NP4","NP3","NP2","NP1")
                               
#Defines a vector for gene names
names_compcoag <- pouch_compcoag$name

#Defines a new data frame consisting of just the CPM columns and call it “pouch_compcoag”, but log-transform (base 2) each CPM value + 0.01.
pouch_compcoag<-log2(pouch_compcoag[,3:14]+0.01)
                             
```

Why do we add 0.01 to all values?

**We add 0.01 in case if a value is close to or at zero due to log of 0 not being defined.**

transpose for row-wise scaling.

```{r}
# Transposes for row-wise scaling.
pouch_compcoag.n <- scale(t(pouch_compcoag))

#Put data back in original orientation
pouch_compcoag.tn <- t(pouch_compcoag.n)

class(pouch_compcoag.tn )
                             
```

What class of object are we dealing with now?

**The class of object we are dealing now is a matrix**

```{r}
#Calculates multivariate dissimilarity for all sample pairs, using Euclidean Distance
compcoag.d1 <- dist(pouch_compcoag.n, method = "euclidean", diag = FALSE, 
                    upper = FALSE)

# Looks at the distances
round(compcoag.d1,3)
                             
```

Which two samples are the most dissimilar based on Euclidean Distance?

**The two samples most dissimilar are NP1 and P3.**


```{r}
#Calculates multivariate dissimilarity for all gene pairs
compcoag.d2 <- dist(pouch_compcoag.tn,method = "euclidean", diag = FALSE, 
                              upper = TRUE)


#Using Ward linkage clustering, cluster samples, then gene
compcoag.c1 <- hclust(compcoag.d1, method = "ward.D2", members = NULL)
compcoag.c2 <- hclust(compcoag.d2, method = "ward.D2", members = NULL)

# Creates a plot with certain parameters.
par(mfrow=c(2,1),cex=0.5) # Make 2 rows, 1 col plot frame and shrink labels
plot(compcoag.c1)
plot(compcoag.c2)
                             
```

What does the sample dendrogram tell us about pregnant (P) and non-pregnant (NP) pouch transcriptomes?

**The sample dendrogram tell us that pregnant pouch transcriptomes are more similar to each other than the non pregnant pouch transcriptomes. Also, the dendogram has 2 major clusters, 1 cluster has all pregnant and the other non-pregnant (NP) pouch transcriptomes.**


```{r}
#Prints the order of samples (left to right) in the tree,
compcoag.c1$order

#Sets the color scale for the heatmap. 299 increments is plenty of resolution!
compcoag_pal <- colorRampPalette(c("green","yellow","red")) 
                                (n = 299)

#Plots the heatmap with heatmap.2 par(cex.main=0.5, cex.axis=0.5, font=2, font.axis=2) # Shrink title fonts on plot

heatmap.2(pouch_compcoag.tn,                       
          Colv=rotate(as.dendrogram(compcoag.c1),  
          order=c(12,11,10,9,8,7,6,5,4,3,2,1)),
          Rowv=as.dendrogram(compcoag.c2),
          labRow=names_compcoag,
          density.info="none",
          trace="none",
          scale="none",            
          col = compcoag_pal,                 
          cexRow=0.5,cexCol=0.75,
          margins=c(3,13), lwid=c(.8,3), lhei=c(.8,3),
          srtCol=45, adjCol=c(1,1),
          keysize=1.3)


                             
```

Are there any groups of genes that differentiate between pregnant and non-pregnant males particularly well? If so, name those genes.

**Yes, the genes that differentiate between pregnant and non-pregnant males particularly well are complement component 6, kinogen, decay accelearating factor and component factor D.**


Repeat everything above
```{r}
# import file into R for analysis
pouch_compcoag_no_scale <- read.delim("CompCoag_pouch_multivar.tsv", sep = "\t", 
                            row.names = 1, header = F)

colnames(pouch_compcoag_no_scale) <- c("KO","name","P9","P8","P7","P6","P5","P3",
                             "NP11","NP10","NP4","NP3","NP2","NP1")

#Defines a vector for gene names
names_compcoag <- pouch_compcoag_no_scale$name

#Defines a new data frame consisting of just the CPM columns, but log-transform (base 2) each CPM value + 0.01.
pouch_compcoag_no_scale<-log2(pouch_compcoag_no_scale[,3:14]+0.01)

# Transposes for row-wise
pouch_compcoag_no_scale.n <- t(pouch_compcoag_no_scale)

#Put data back in original orientation
pouch_compcoag_no_scale.tn <- t(pouch_compcoag_no_scale.n)

class(pouch_compcoag_no_scale.tn)


#Calculates multivariate dissimilarity for all sample pairs, using Euclidean Distance
compcoag.d1_no_scale <- dist(pouch_compcoag_no_scale.n, method = "euclidean", diag = FALSE, 
                   upper = FALSE)

# Looks at the distances
round(compcoag.d1_no_scale,3)

#Calculates multivariate dissimilarity for all gene pairs
compcoag.d2_no_scale <- dist(pouch_compcoag_no_scale.tn,method = "euclidean", diag = FALSE, 
                   upper = TRUE)

#Using Ward linkage clustering, cluster samples, then gene
compcoag.c1_no_scale <- hclust(compcoag.d1_no_scale, method = "ward.D2", members = NULL)
compcoag.c2_no_scale <- hclust(compcoag.d2_no_scale, method = "ward.D2", members = NULL)

# Creates a plot with certain parameters
par(mfrow=c(2,1),cex=0.5) # Make 2 rows, 1 col plot frame and shrink labels
plot(compcoag.c1_no_scale)
plot(compcoag.c2_no_scale)

#Prints the order of samples (left to right) in the tree
compcoag.c1_no_scale$order

#Sets the color scale for the heatmap. 299 increments is plenty of resolution!
compcoag_pal <- colorRampPalette(c("green","yellow","red")) 
(n = 299)


#Plots the heatmap with heatmap.2 par(cex.main=0.5, cex.axis=0.5, font=2, font.axis=2) # Shrink title fonts on plot
heatmap.2(pouch_compcoag_no_scale.tn,                       
         Colv=rotate(as.dendrogram(compcoag.c1_no_scale),  
                     order=c(12,11,10,9,8,7,6,5,4,3,2,1)),
         Rowv=as.dendrogram(compcoag.c2_no_scale),
         labRow=names_compcoag,
         density.info="none",
         trace="none",
         scale="none",            
         col = compcoag_pal,                 
         cexRow=0.5,cexCol=0.75,
         margins=c(3,13), lwid=c(.8,3), lhei=c(.8,3),
         srtCol=45, adjCol=c(1,1),
         keysize=1.3)
                             
```
 
 Does your heatmap look any different? If so, explain why this might be.
 
**The heatmap looks different. There seems to be more upregulated expression, extreme values, throughout the heat map and there is no particularly well differentiation between pregnant and non-pregnant males. The scaling might have an effect on the graphic output by scaling such that the color choice is improved, so the color choice would not be dominated by the extreme values Also, this could be because not having the scale function increases the stress values.**


```{r}
# import file into R for analysis
subsample <- read.delim("stickleback_CPM.tsv", sep = "\t", 
                             row.names = 1, header = T)

#subsets only genes that have the Genome_Loc value “groupXIX” AND have a Gene_Start_Pos value between 6000000 and 12000000.
subsample<-subset(subsample, subsample$Genome_Loc=="groupXIX" & subsample$Gene_Start_Pos>6000000 & subsample$Gene_Start_Pos<12000000)

#Defines a new data frame consisting of just the CPM column, but log-transform (base 2) each CPM value + 0.01.
stickle_CPM<-log2(subsample[,3:length(colnames(subsample))]+0.01)

# Transposes for row-wise scaling.
subsample.n<-scale(t(stickle_CPM))

#Put data back in original orientation
subsample.tn <- t(subsample.n)

#Calculates multivariate dissimilarity for all sample pairs, using Euclidean Distance
subsample.d1 <- dist(subsample.n, method = "euclidean", diag = FALSE, 
                    upper = FALSE)
# Looks at the distances
#round(subsample.d1,3)

#Calculates multivariate dissimilarity for all gene pairs
subsample.d2 <- dist(subsample.tn,method = "euclidean", diag = FALSE, 
                    upper = TRUE)

#Using Ward linkage clustering, cluster samples, then gene
subsample.c1 <- hclust(subsample.d1, method = "ward.D2", members = NULL)
subsample.c2 <- hclust(subsample.d2, method = "ward.D2", members = NULL)

# Creates a plot with certain parameters.
par(mfrow=c(2,1),cex=0.5) # Make 2 rows, 1 col plot frame and shrink labels
plot(subsample.c1)
plot(subsample.c2)
                    

#Prints the order of samples (left to right) in the tree
subsample.c1$order

#Sets the color scale for the heatmap. 299 increments is plenty of resolution!
subsample_pal <- colorRampPalette(c("green","yellow","red"))(n = 299)

#Plots the heatmap with heatmap.2 par(cex.main=0.5, cex.axis=0.5, font=2, font.axis=2) # Shrink title fonts on plot
heatmap.2(subsample.tn,                       
          Colv=rotate(as.dendrogram(subsample.c1 ),  
                      order=c(12,11,10,9,8,7,6,5,4,3,2,1)),
          Rowv=as.dendrogram(subsample.c2),
          #labRow=names_subsample,
          density.info="none",
          trace="none",
          scale="none",            
          col = subsample_pal,                 
          cexRow=0.5,cexCol=0.75,
          margins=c(3,13), lwid=c(.8,3), lhei=c(.8,3),
          srtCol=45, adjCol=c(1,1),
          keysize=1.3)
```


```{r}

# import file into R for analysis
subsample <- read.delim("stickleback_CPM.tsv", sep = "\t", 
                        row.names = 1, header = T)


#Defines a new data frame consisting of just the CPM columns, but log-transform (base 2) each CPM value + 0.01.
stickle_CPM<-log2(subsample[,3:length(colnames(subsample))]+0.01)

#Takes a random sample of genes the same size as your first subset.
subsample2 <- stickle_CPM[sample(nrow(stickle_CPM),314),]

# Transposes for row-wise scaling.
subsample2.n<-scale(t(subsample2))

#Put data back in original orientation
subsample2.tn <- t(subsample2.n)

#Calculates multivariate dissimilarity for all sample pairs, using Euclidean Distance
subsample2.d1 <- dist(subsample2.n, method = "euclidean", diag = FALSE, 
                     upper = FALSE)
# Looks at the distances
#round(subsample2.d1,3)

#Calculates multivariate dissimilarity for all gene pairs
subsample2.d2 <- dist(subsample2.tn,method = "euclidean", diag = FALSE, 
                     upper = TRUE)

#Using Ward linkage clustering, cluster samples, then gene
subsample2.c1 <- hclust(subsample2.d1, method = "ward.D2", members = NULL)
subsample2.c2 <- hclust(subsample2.d2, method = "ward.D2", members = NULL)


# Creates a plot with certain parameters.
par(mfrow=c(2,1),cex=0.5) # Make 2 rows, 1 col plot frame and shrink labels
plot(subsample.c1)
plot(subsample.c2)

#Prints the order of samples (left to right) in the tree
subsample2.c1$order

#Sets the color scale for the heatmap. 299 increments is plenty of resolution!
subsample2_pal <- colorRampPalette(c("green","yellow","red"))(n = 299)

#Plots the heatmap with heatmap.2 par(cex.main=0.5, cex.axis=0.5, font=2, font.axis=2) # Shrink title fonts on plot
heatmap.2(subsample2.tn,                       
          Colv=rotate(as.dendrogram(subsample2.c1 ),  
                      order=c(12,11,10,9,8,7,6,5,4,3,2,1)),
          Rowv=as.dendrogram(subsample2.c2),
          #labRow=names_subsample,
          density.info="none",
          trace="none",
          scale="none",            
          col = subsample2_pal,                 
          cexRow=0.5,cexCol=0.75,
          margins=c(3,13), lwid=c(.8,3), lhei=c(.8,3),
          srtCol=45, adjCol=c(1,1),
          keysize=1.3)

                             
```


Do you notice anything different about the clustering patterns for males and females between the two subsets? Do a quick literature search to find out why we might expect a difference given the biology of stickleback chromosome 19, and interpret what may be going on, especially regarding the clusters of genes that yield different sex-specific expression patterns in the groupXIX heatmap.

**There was a difference, the first heatmap focused on chromosome 19 which is a sex determine locus. controls male or female sexual development in threespine sticklebacks. The dendogram that is oriented on top of plot, we see two distinct branches. Seems like the branch on the left is male and the other branch on the right is female. The branch on the right, which seems to be female, has majority upregulated gene expression compared to male, branch on the left, that shows majority downregulated gene expression. The second heat map was a random samples chromosome positions and did not show any pattern of upregulated or downregulated expression to being male or female.**








